source/dhtslib/vcf/header.d

```diff
@@ -1,8 +1,260 @@
+    void setLength(T)(T number)
+    if(isIntegral!T)
+    {
+        this.lenthType = HeaderLengths.FIXED;
+        this["Number"] = number.to!string;
+    }
+
+    /// set Value Type length i.e A, R, G, .
+    void setLength(HeaderLengths number)
+    {
+        this.lenthType = number;
+        this["Number"] = HeaderLengthsStrings[number];
+    }
+
+    /// get Value Type length
+    string getLength()
+    {
+        return this["Number"];
+    }
+
+    /// set Value Type i.e Integer, String, Float
+    void setValueType(HeaderTypes type)
+    {
+        this.valueType = type;
+        this["Type"] = HeaderTypesStrings[type];
+    }
+
+    /// get Value Type i.e Integer, String, Float
+    HeaderTypes getValueType()
+    {
+        return this.valueType;
+    }
+
+    /// set ID field
+    void setID(string id)
+    {
+        this["ID"] = id;
+    }
+
+    /// get ID field
+    string getID()
+    {
+        return this["ID"];
+    }
+
+    /// set Description field
+    void setDescription(string des)
+    {
+        this["Description"] = des;
+    }
+
+    /// get Description field
+    string getDescription()
+    {
+        return this["Description"];
+    }
+
+    /// get a value from the KV pairs
+    /// if key isn't present thows exception
+    ref auto opIndex(string index)
+    {
+        foreach (i, string key; keys)
+        {
+            if(key == index){
+                return vals[i];
+            }
+        }
+        throw new Exception("Key " ~ index ~" not found");
+    }
+
+    /// set a value from the KV pairs
+    /// if key isn't present a new KV pair is 
+    /// added
+    void opIndexAssign(string value, string index)
+    {
+        foreach (i, string key; keys)
+        {
+            if(key == index){
+                vals[i] = value;
+                return;
+            }
+        }
+        this.nkeys++;
+        keys~=index;
+        vals~=value;
+    }
+
+    /// convert to bcf_hrec_t for use with htslib functions
+    bcf_hrec_t * convert(bcf_hdr_t * hdr)
+    {
+        if(this.recType == HeaderRecordType.INFO || this.recType == HeaderRecordType.FORMAT){
+            assert(this.valueType != HeaderTypes.NULL);
+            assert(this.lenthType != HeaderLengths.NULL);
+        }
+
+        auto str = this.toString;
+        int parsed;
+        auto rec = bcf_hdr_parse_line(hdr, toUTFz!(char *)(str), &parsed);
+        rec.type = this.recType;
+        return rec;
+    }
+
+    /// print a string representation of the header record
+    string toString()
+    {
+        string ret = "##" ~ this.key ~ "=" ~ this.value;
+        if(this.nkeys > 0){
+            ret ~= "<";
+            for(auto i =0; i < this.nkeys - 1; i++)
+            {
+                ret ~= this.keys[i] ~ "=" ~ this.vals[i] ~ ", ";
+            }
+            ret ~= this.keys[$-1] ~ "=" ~ this.vals[$-1] ~ ">";    
+        }
+        
+        return ret;
+    }
+}
 
 /** Wrapper around bcf_hdr_t
 

```

Comment:
since we do have SAMHeader this doc comment should be revised

Reply:
Agreed

Reply:
addressed in
1056615